main/src/library/Map.flix

```diff
@@ -1,7 +1,335 @@
+
+    ///
+    /// Returns the values of `m`.
+    ///
+    def valuesOf[k, v](m: Map[k, v]): List[v] =
+        let Map(xs) = m;
+            snd(List.unzip(xs))
+
+    //
+    // ## Insert
+    //
+
+    ///
+    /// Updates `m` with `k -> v`.
+    ///
+    def insert[k, v](k: k, v: v, m: Map[k, v]): Map[k, v] =
+        insertWithKey((k1, v1, v2) -> v1, k, v, m)
+
+    ///
+    /// Updates `m` with `k -> f(v, v1)` if `k -> v1` is in `m`.
+    /// Otherwise, updates `m` with `k -> v`.
+    ///
+    def insertWith[k, v](f: (v, v) -> v, k: k, v: v, m: Map[k, v]): Map[k, v] =
+        insertWithKey((k1, v1, v2) -> f(v1, v2), k, v, m)
+
+    ///
+    /// Updates `m` with `k -> f(k, v, v1)` if `k -> v1` is in `m`.
+    /// Otherwise, updates `m` with `k -> v`.
+    ///
+    def insertWithKey[k, v](f: (k, v, v) -> v, k: k, v: v, m: Map[k, v]): Map[k, v] =
+        let Map(xs) = m;
+            Map(insertWithKeyHelper(f, k, v, xs))
+
+    ///
+    /// Helper function for `insertWithKey`.
+    ///
+    @internal
+    def insertWithKeyHelper[k, v](f: (k, v, v) -> v, k: k, v: v, xs: List[(k, v)]): List[(k, v)] = match xs with {
+        case Nil => (k, v) :: Nil
+        case (a, b) :: rs => if (k == a) (k, f(k, v, b)) :: rs else (a, b) :: insertWithKeyHelper(f, k, v, rs)
+    }
+
+    //
+    // ## Update
+    //
+
+    ///
+    /// Updates `m` with `k -> f(v)` if `k -> v` is in `m`.
+    /// Otherwise, returns `m`.
+    ///
+    def adjust[k, v](f: v -> v, k: k, m: Map[k, v]): Map[k, v] = adjustWithKey((k1, v1) -> f(v1), k, m)
+
+    ///
+    /// Updates `m` with `k -> f(k, v)` if `k -> v` is in `m`.
+    /// Otherwise, returns `m`.
+    ///
+    def adjustWithKey[k, v](f: (k, v) -> v, k: k, m: Map[k, v]): Map[k, v] =
+        let Map(xs) = m;
+            Map(List.map(match (key, v) -> if (key == k) (k, f(k, v)) else (key, v), xs))
+
+    ///
+    /// Updates `m` with `k -> v1` if `k -> v` is in `m` and `f(v) = Some(v1).
+    /// Otherwise, returns `m`.
+    ///
+    def update[k, v](f: v -> Option[v], k: k, m: Map[k, v]): Map[k, v] = updateWithKey((k1, v1) -> f(v1), k, m)
+
+    ///
+    /// Updates `m` with `k -> v1` if `k -> v` is in `m` and `f(k, v) = Some(v1).
+    /// Otherwise, returns `m`.
+    ///
+    def updateWithKey[k, v](f: (k, v) -> Option[v], k: k, m: Map[k, v]): Map[k, v] =
+        let Map(xs) = m;
+            Map(List.map(match (key, v) -> if (key == k) (k, Option.getWithDefault(f(k, v), v)) else (key, v), xs))
+
+    //
+    // ## Delete
+    //
+
+    ///
+    ///
+    ///
+    def delete[k, v](k: k, m: Map[k, v]): Map[k, v] =
+        let Map(xs) = m;
+            Map(List.filter(match (key, v) -> key != k, xs))
+
+    //
+    // ## Map Predicates
+    //
+
+    ///
+    /// Returns `true` if and only if all mappings in `m1` occur in `m2`.
+    ///
+    def isSubmapOf[k, v](m1: Map[k, v], m2: Map[k, v]): Bool =
+        let Map(xs) = m1;
+        let Map(ys) = m2;
+            List.forall(x -> List.memberOf(x, ys), xs)
```

Comment:
Here `x` is actually a pair (right?), so I would prefer it be called `kv` to be clear it is both the key and value